incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,48 +64,70 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
-		return deny(metav1.StatusReasonForbidden, msg)
-	}
-
-	return v.checkForest(req)
-}
-
-func (v *Namespace) checkForest(req *nsRequest) admission.Response {
 	v.Forest.Lock()
 	defer v.Forest.Unlock()
 
 	ns := v.Forest.Get(req.ns.Name)
 
-	// Allow all create namespace requests except if the namespace name already exists in
-	// the external hierarchy (the apiserver already denies existing namespace names).
-	if req.op == v1beta1.Create {
-		for _, nm := range v.Forest.GetNamespaceNames() {
-			if _, ok := v.Forest.Get(nm).ExternalTreeLabels[req.ns.Name]; ok {
-				msg := fmt.Sprintf("The namespace name %q is reserved by the external hierarchy manager %q.", req.ns.Name, v.Forest.Get(nm).Manager)
-				return deny(metav1.StatusReasonAlreadyExists, msg)
-			}
+	switch req.op {
+	case v1beta1.Create:
+		// Allow all create namespace requests except if the namespace name already exists in
+		// the external hierarchy (the apiserver already denies existing namespace names).
+		return v.nameExistsInExternalHierarchy(req)
+	case v1beta1.Update:
+		// Allow all update namespace requests except if the namespace has a parent and wants
+		// to add "hnc.x-k8s.io/managedBy" annotation other than "hnc.x-k8s.io".
+		return v.conflictBetweenParentAndExternalManager(req, ns)
+	case v1beta1.Delete:
+		if rsp := v.cannotDeleteSubnamespace(req); !rsp.Allowed {
+			return rsp
 		}
+		return v.illegalCascadingDeletion(ns)
 	}
 
-	// Allow all update namespace requests except if the namespace has a parent and wants
-	// to add "hnc.x-k8s.io/managedBy" annotation other than "hnc.x-k8s.io".
-	if req.op == v1beta1.Update && req.ns.Annotations[api.AnnotationManagedBy] != "" {
-		if req.ns.Annotations[api.AnnotationManagedBy] != api.MetaGroup && ns.Parent() != nil {
-			msg := fmt.Sprintf("Namespace %q is a child of %q. Namespaces with parents defined by HNC cannot also be managed externally. "+
-				"To manage this namespace with %q, first make it a root in HNC.", ns.Name(), ns.Parent().Name(), req.ns.Annotations[api.AnnotationManagedBy])
-			return deny(metav1.StatusReasonForbidden, msg)
+	return allow("")
+}
+
+// nameExistsInExternalHierarchy only applies to the Create operation since
+// namespace name cannot be updated.
+func (v *Namespace) nameExistsInExternalHierarchy(req *nsRequest) admission.Response {
+	for _, nm := range v.Forest.GetNamespaceNames() {
+		if _, ok := v.Forest.Get(nm).ExternalTreeLabels[req.ns.Name]; ok {
+			msg := fmt.Sprintf("The namespace name %q is reserved by the external hierarchy manager %q.", req.ns.Name, v.Forest.Get(nm).Manager)
+			return deny(metav1.StatusReasonAlreadyExists, msg)
 		}
 	}
+	return allow("")
+}
 
-	// Early exit to allow other non-delete requests or if the namespace allows cascading deletion.
-	if req.op != v1beta1.Delete || ns.AllowsCascadingDelete() {
+// conflictBetweenParentAndExternalManager only applies to the Update operation
+// to check if changing namespace manager is allowed.
```

Comment:
Ditto (this comment belongs in the switch statement, IMO). Maybe explain _why_ this can't be a problem with Create.